mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -0,0 +1,422 @@
+                if_match,
+                part_queue_producer,
+            );
+            let span = debug_span!("prefetch", ?range);
+            async move {
+                request.get_from_cache(range).await;
+            }
+            .instrument(span)
+        };
+
+        let task_handle = self.runtime.spawn_with_handle(request_task).unwrap();
+
+        RequestTask::from_handle(task_handle, size, start, part_queue)
+    }
+}
+
+fn get_aligned_request_range(range: RequestRange, block_size: u64) -> RequestRange {
+    let object_size = range.object_size();
+    let offset = range.start();
+    let preferred_size = range.len();
+
+    // If the request size is bigger than a block size we will try to align it to block boundaries.
+    let offset_in_part = offset % block_size;
+    let size = if offset_in_part != 0 {
+        // if the offset is not at the start of the part we will drain all the bytes from that part first
+        let remaining_in_part = block_size - offset_in_part;
+        preferred_size.min(remaining_in_part as usize)
+    } else {
+        // if the request size is smaller than the block size, just return the block size
+        if preferred_size < block_size as usize {
+            block_size as usize
+        } else {
+            // if it exceeds block boundaries, trim it to the block boundaries
+            let request_boundary = offset + preferred_size as u64;
+            let remainder = request_boundary % block_size;
+            if remainder != 0 {
+                preferred_size + (block_size - remainder) as usize
+            } else {
+                preferred_size
+            }
+        }
+    };
+    RequestRange::new(object_size, offset, size)
+}
+
+#[derive(Debug)]
+struct CachingRequest<Client: ObjectClient, Cache> {
+    client: Client,
+    cache: Arc<Cache>,
+    bucket: String,
+    cache_key: CacheKey,
+    part_queue_producer: PartQueueProducer<Client::ClientError>,
+}
+
+impl<Client, Cache> CachingRequest<Client, Cache>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+    Cache: DataCache + Send + Sync,
+{
+    fn new(
+        client: Client,
+        cache: Arc<Cache>,
+        bucket: String,
+        key: String,
+        etag: ETag,
+        part_queue_producer: PartQueueProducer<Client::ClientError>,
+    ) -> Self {
+        let cache_key = CacheKey { s3_key: key, etag };
+        Self {
+            client,
+            cache,
+            bucket,
+            cache_key,
+            part_queue_producer,
+        }
+    }
+
+    async fn get_from_cache(&self, range: RequestRange) {
+        let key = &self.cache_key.s3_key;
+        let block_size = self.cache.block_size();
+        let block_range = self.block_indices_for_byte_range(&range);
+
+        // TODO: consider starting GetObject requests pre-emptively if cache blocks are missing
+        for block_index in block_range.clone() {
+            match self.cache.get_block(&self.cache_key, block_index) {
```

Comment:
Not for this commit: we'll want some metrics on cache hit/miss latency and cache hit rate.